README.md
# socket-json-rpc

[![Build Status](https://travis-ci.org/invlpg/socket-json-rpc.svg?branch=master)](https://travis-ci.org/invlpg/socket-json-rpc)
[![codecov](https://codecov.io/gh/invlpg/socket-json-rpc/branch/master/graph/badge.svg)](https://codecov.io/gh/invlpg/socket-json-rpc)
[![Dependencies Status](https://jarkeeper.com/invlpg/socket-json-rpc/status.svg)](https://jarkeeper.com/invlpg/socket-json-rpc)

[![Clojars Project](https://img.shields.io/clojars/v/invlpg/socket-json-rpc.svg)](https://clojars.org/invlpg/socket-json-rpc)

socket-json-rpc is a Clojure library that simplifies creating async JSON-RPC
servers and clients that communicate using TCP sockets.

## Server Usage

First, require the server:

```clojure
(require '[socket-json-rpc.server :as server])
```

Procedures are defined with `defprocedure`, and is used in the form

```clojure
(server/defprocedure [name] [[named-args]] [body])
```

The named-args argument tells the server how to arrange named arguments into a
vector so that the body can use the arguments in a consistent manner. For
example, the following procedure allows both unnamed and named argument usage.

```clojure
(server/defprocedure subtract
  ["minuend" "subtrahend"]
  (server/respond (- (first args) (second args))))
```

*Note:* The body automatically receives the variables `args` and `notification`.
`args` is a vector of supplied arguments, matching the order specified in the
`named-args` vector supplied when the procedure was defined. `notification` is a
boolean value that is true when the client does not want to receive a reply with
the result of the procedure call. If the procedure does not respect the
`notification` variable, the server will still not send the result anyway.

*Note:* Normally, the number of unnamed arguments supplied must match the number
of named arguments supplied in the procedure definition, but to allow for
variable arguments, the special named argument `"..."` can be used. For example:

```clojure
(server/defprocedure add
  ["a" "..."]
  (server/respond (reduce + args)))
```

The client must still provide at least the number of named arguments other than
`"..."`. In the previous example, at least one number would have to be specified.

The `subtract` method can be called with either of the following JSON:

```javascript
// Unnamed arguments
{
  "jsonrpc": "2.0",
  "method": "subtract",
  "params": [42, 23],
  "id": 1
}
```

```javascript
// Named arguments
{
  "jsonrpc": "2.0",
  "method": "subtract",
  "params": {
    "subtrahend": 23,
    "minuend": 42
  },
  "id": 1
}
```

And both would return the same result:

```javascript
{
  "jsonrpc": "2.0",
  "result": "19",
  "id": 1
}
```

The server can then be started in one of two ways:

```clojure
; This version blocks the calling thread, so it won't close.
(defn -main
  [& args]
  ...
  (server/start [port] [backlog] [bind-addr]))
```

```clojure
; This version allows the calling thread to continue immediately, meaning that
; if the main thread closes, the server will also be shut down.
(defn -main
  [& args]
  ...
  (server/start-async [port] [backlog] [bind-addr])
  ...)
```

The only required argument to either form is `port`. `backlog` specifies how many
pending connections should be buffered, and is 50 by default. `bind-addr`
specifies the host address that the server should listen on.

## Client Usage

First, require the client:

```clojure
(require '[socket-json-rpc.client :as client :refer (notify)])
```

The client is very simple, only exposing one macro - `with-server`, and is used
in the form

```clojure
(client/with-server [host] [port] [body])
```

This form allows the client to call functions on the server *almost* as if they
were local functions. For example, our `subtract` procedure defined above could
be called on a local server running on port `9001` by either:

Using the ordered argument form:

```clojure
(client/with-server "localhost" 9001
  (subtract 42 23))
```

Or, using the named argument form:

```clojure
(client/with-server "localhost" 9001
  (subtract :subtrahend 23 :minuend 42))
```

Both forms would return the same result:

```clojure
[(false 19)]
```

`with-server` returns a vector of function results. Each item in the vector is a
list containing

```clojure
'([error] 
   [result] || [code] [message])
  
; i.e. successful function call
'(false 19)

; i.e. error condition
'(true -32601 "Method not found")
```

A remote procedure call can be wrapped in `notify` if you aren't interested in
receiving a reply with the result of the function call.

Multiple functions can be grouped together into a single "batch" by calling them
successively inside the `with-server` block. Note that each function will be
called on the server in-order.

```clojure
(client/with-server "localhost" 9001
  (subtract 42 23)
  (add 1 2 3 4 5)
  (subtract :subtrahend 42 :minuend 23)
  (notify (add 15 15))
  (doesntexist true))
  
; => [(false 19)
;     (false 15)
;     (false -19)
;     (true -32601 "Method not found")]
```

Notice that the return values arrive in the same order as the calls were
delivered. 

## Common Problems

```clojure
ClassCastException java.lang.String cannot be cast to clojure.lang.Associative
```

This usually means that you forgot to add `(respond [val])` around your procedure's
return value.

## License

Copyright © 2017 Carl Albrecht

Distributed under the Eclipse Public License version 1.0


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/), and uses [Semantic Versioning](http://semver.org).

## [Unreleased]
### Changed

## [1.0.0] - 2017-03-23
### Added
- `start` and `start-async` functions to serve incoming socket RPC requests
- `defprocedure` to create new procedures that can then be called remotely
- `respond` and `error` for server-side procedures to generate responses
- `socket-io` private function to handle reading and writing to server sockets
- `execute` private function to parse the request and decide whether the request
  is a single procedure call or a batch
- `do-single` and `do-batch` to separate individual calls, and execute, before
  arranging return JSON
- `execute-single` to execute the map representation of a single procedure call,
  if it is valid
- `when-valid` executes its body if the input to `execute-single` is in a valid
  JSON-RPC request format
- `named-params` and `positional-params` organise input parameters into a
  consistent format, before calling the procedure defined by `defprocedure`
- `with-server` macro for clients to call procedures on a server
- `notify` can be wrapped around a procedure call if a client does not want a
  response
- `socket-io` to send a request to the specified server, then read the response,
  organising responses by `id` in the case of a batch.
- `read-all` handles reading the character count at the top of the response, then
  reading an appropriate number of characters based on that count
- `parse-body` converts the body of a `with-server` block from clojure function
  form to JSON-RPC form, checking for validity
- README with basic API documentation

[Unreleased]: https://github.com/invlpg/socket-json-rpc/compare/v1.0.0...HEAD


test/socket_json_rpc/core_test.clj
(ns socket-json-rpc.core-test
  (:require [clojure.test :refer :all]
            [socket-json-rpc.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/socket_json_rpc/comms_test.clj
(ns socket-json-rpc.comms-test
  (:require  [clojure.test :refer :all]
             [clojure.data.json :as json]
             [socket-json-rpc.server :as server]
             [socket-json-rpc.client :as client :refer (notify)])
  (:import [java.net InetAddress]))

; TODO figure out why notify only works when using the full path
; i.e. socket-json-rpc.client/notify instead of just notify or client/notify
; n.b. This issue only seems to happen inside tests :\
(deftest communication-async
  ; All of these servers should work the same...
  (println "wasd")
  (server/start-async 34575)
  (println "esdf")
  (future (server/start 34576))

  (server/defprocedure subtract
    ["minuend" "subtrahend"]
    (server/respond (- (first args) (second args))))

  (server/defprocedure add
    ["..."]
    (server/respond (reduce + args)))


  (testing "single call unnamed args"
    (is (= (client/with-server "localhost" 34575
             (subtract 42 23))
           '[(false 19)])))


  (testing "single call named args"
    (is (= (client/with-server "localhost" 34575
             (subtract :subtrahend 42 :minuend 23))
           '[(false -19)])))


  (testing "single call notification"
    (is (= (client/with-server "localhost" 34575
             (socket-json-rpc.client/notify (subtract 42 23)))
           nil)))


  (testing "single call error"
    (is (= (client/with-server "localhost" 34576
             (non-existant true))
           '[(true -32601 "Method not found")])))


  (alter-var-root #'server/zero-string (constantly "0\n"))

  (testing "single call notify error"
    (is (= (client/with-server "localhost" 34576
             (socket-json-rpc.client/notify (non-existant true)))
           nil)))


  (testing "batch call"
    (is (= (client/with-server "localhost" 34576
             (subtract :minuend 512 :subtrahend 256)
             (socket-json-rpc.client/notify (subtract 5 2))
             (add 1 2 3 4 5)
             (non-existant true)
             (subtract 5)
             (socket-json-rpc.client/notify (non-existant true))
             (add 3))
           '[(false 256)
             (false 15)
             (true -32601 "Method not found")
             (true -32602 "Invalid params")
             (false 3)]))))




test/socket_json_rpc/client_test.clj
(ns socket-json-rpc.client-test
  (:require [clojure.test :refer :all]
            [clojure.string :as str]
            [clojure.data.json :as json]
            [socket-json-rpc.client :as client]
            [socket-json-rpc.server :as server]))

(deftest client-errors
  (testing "keyword arguments uneven number of forms"
    (is (thrown? IllegalArgumentException
                 (client/with-server "localhost" 12345
                   (test :hi))))))


; Force server to generate multi-line responses
(ns socket-json-rpc.server)
(defn set-use-newlines [f]
  (def execute-orig execute)

  (defn- execute
    [request]
    (let [result (execute-orig request)]
      (str/replace result #"," ",\n")))

  (f))
(ns socket-json-rpc.client-test)
(use-fixtures :once server/set-use-newlines)


(deftest client-misc
  (server/defprocedure subtract
    ["minuend" "subtrahend"]
    (server/respond (- (first args) (second args))))

  (println "wasd")

  (server/start-async 34577)

  (testing "json spread over multiple lines"
    (is (= (client/with-server "localhost" 34577
             (subtract 1337 420))
           '[(false 917)]))))



src/socket_json_rpc/client.clj
(ns socket-json-rpc.client
  (:require [clojure.data.json :as json]
            [clojure.core.async :as async]
            [com.gearswithingears.async-sockets :refer :all]))

(defn- parse-body
  "Converts function call(s) from clojure s-expression form to JSON RPC form"
  [body]
  (loop [body body out nil id 1]
    (if-let [func (if (list? (second (first body)))
                    (second (first body))
                    (first body))]
      (let [method (str (first func))
            args (rest func)]
        (let [args (if (keyword? (first args))
                     (if (odd? (count args))
                       (throw (new IllegalArgumentException "Keyword arguments must contain an even number of forms"))
                       (zipmap (take-nth 2 args) (take-nth 2 (rest args))))
                     args)]
          (let [request {"jsonrpc" "2.0"
                         "method" method
                         "params" args
                         "id" id}]
            (recur (rest body)
                   (conj out (if (list? (second (first body)))
                               (@(resolve (first (first body))) request)
                               request))
                   (inc id)))))
      (if (= (count out) 1)
        (first out)
        out))))

(defn notify
  "Filters out the 'id' key from a single RPC request."
  [request]
  (dissoc request "id"))

; (Should probably create a common file for all of these functions ;)
; From http://stackoverflow.com/a/12503724
(defn- parse-int
  "Parses the first continuous number only"
  [s]
  (new Integer (re-find #"\d+" s)))

(defn- read-all
  "Reads the character count on the first line of the response and then reads
  specified number of characters from the response, then returns the response
  body."
  [socket]
  (loop [length -1 response ""]
    (when-let [line (async/<!! (:in socket))]
      (if (or (nil? length) (zero? length))
        nil
        (if (neg? length)
          (recur (try (parse-int line)
                      (catch Exception _ nil))
                 response)
          (let [response (str response line)]
            (if (< (count response) length)
              (recur length (str response "\n"))
              response)))))))

(defn- socket-io
  "Sends the collated JSON-RPC request to the RPC server, and then returns a
  properly formatted response."
  [address port out]
  (let [socket (socket-client port address)]
    (let [jsonout (json/write-str out)]
      (async/>!! (:out socket) (str (count jsonout) "\n" jsonout))
      (if-let [ret (read-all socket)]
        (do
          (close-socket-client socket)
          (let [return (json/read-str ret)]
            (if (map? return)
              (if (contains? return "error")
                (vector (list true (get-in return ["error" "code"]) (get-in return ["error" "message"])))
                (vector (list false (get return "result"))))
              (loop [return return result []]
                (if-let [response (first return)]
                  (if (contains? response "error")
                    (recur (rest return)
                           (conj result
                                 (list (get response "id")
                                       true
                                       (get-in response ["error" "code"])
                                       (get-in response ["error" "message"]))))
                    (recur (rest return)
                           (conj result
                                 (list (get response "id")
                                       false
                                       (get response "result")))))
                  (vec (map rest (sort-by first result))))))))
        (do
          (close-socket-client socket)
          nil)))))

(defmacro with-server
  "Runs the body on a specified server. Ordered parameters are supplied just
  like supplying args to any other Clojure function. Named parameters are
  supplied by using keyword arguments."
  [address port & body]
  `(when-let [out# (#'parse-body '~body)]
     (#'socket-io ~address ~port out#)))


src/socket_json_rpc/server.clj
(ns socket-json-rpc.server
  (:require [clojure.string :as str]
            [clojure.data.json :as json]
            [clojure.core.async :as async]
            [com.gearswithingears.async-sockets :refer :all])
  (:import [java.net InetAddress]))

(defmacro def-
  "Define a new private variable in 'defn-' style."
  [sym init]
  `(def ~(with-meta sym {:private true}) ~init))

(def- jrpc-namespace {})

(def- parse-error '(-32700 "Parse error"))
(def- invalid-request '(-32600 "Invalid Request"))
(def- method-not-found '(-32601 "Method not found"))
(def- invalid-params '(-32602 "Invalid params"))
(def- internal-error '(-32603 "Internal error"))

(def- zero-string "\n")

(defn error
  "Generates the return JSON for an error condition."
  ([err]
    (error (first err) (second err)))
  ([code message]
    (if (and (<= code -32000) (>= code -32768))
      {"jsonrpc" "2.0"
       "error" {"code" -32000
                "message" "Application attempted to use reserved error code"}}
      {"jsonrpc" "2.0"
       "error" {"code" code
                "message" message}})))

(defn- server-error
  "Generates the return JSON for an internal server error."
  ([err]
    (server-error (first err) (second err)))
  ([code message]
    {"jsonrpc" "2.0"
     "error" {"code" code
              "message" message}}))

(defn respond
  "Generates the return JSON for a function's return value."
  [result]
  {"jsonrpc" "2.0"
   "result" result})

(defmacro when-valid
  "Continues with body if the RPC call is valid. Otherwise, returns an appropriate
  error. body automatically receives the variables 'params', 'id' and 'method'"
  [call body]
  `(let [params# (get ~call "params")
         id# (get ~call "id")]
     (if (= (get ~call "jsonrpc") "2.0")
       (if (string? (get ~call "method"))
         (if-let [method# (get (var-get #'jrpc-namespace) (get ~call "method"))]
           (if (or (vector? params#) (map? params#) (nil? params#))
             ((fn [~'params ~'id ~'method] ~body) params# id# method#)
             (assoc (#'server-error (var-get #'invalid-request)) "id" id#))
           (assoc (#'server-error (var-get #'method-not-found)) "id" id#))
         (assoc (#'server-error (var-get #'invalid-request)) "id" id#))
       (assoc (#'server-error (var-get #'invalid-request)) "id" id#))))

(defmacro named-params
  "Organises named parameters into a vector as if it were a positional parameter
  call, then calls the procedure as per normal."
  [call params id method]
  `(loop [args# (into [] (map (constantly nil) (first ~method)))
          params# ~params]
     (if-not (vector? args#)
       args#
       (if-let [arg# (first params#)]
         (recur (try (assoc args# (.indexOf (first ~method) (first arg#)) (second arg#))
                     (catch Exception _#
                       (assoc (#'server-error (var-get #'invalid-params)) "id" ~id)))
                (rest params#))
         (assoc ((second ~method) args# (not (contains? ~call "id"))) "id" ~id)))))

(defmacro positional-params
  "Just passes the vector of arguments to the procedure as-is."
  [call params id method]
  `(if (nil? ~params)
     (assoc ((second ~method) nil (not (contains? ~call "id"))) "id" ~id)
     (if (or (= (count (first ~method)) (count ~params))
             (and (= (last (first ~method)) "...")
                  (<= (count (first ~method)) (count ~params))))
       ;(do (println ~params)
       (assoc ((second ~method) ~params (not (contains? ~call "id"))) "id" ~id) ;)
       (assoc (#'server-error (var-get #'invalid-params)) "id" ~id))))

(defn- execute-single
  "Takes the JSON input of a single procedure call, and returns a map containing
  the return JSON, if applicable (i.e. notifications do not return anything).
  Handles adding ids to return values automatically."
  [call]
  (when-valid call
    (if (map? params)
      (named-params call params id method)
      (positional-params call params id method))))

(defmacro do-batch
  "Organises the execution of each entry in a batch, as well as the collation and
  return of the outputs of each entry. Enforces not returning anything if the call
  is a notification."
  [input]
  `(if (empty? ~input)
     (json/write-str (assoc (#'server-error (var-get #'invalid-request)) "id" nil))
     (loop [input# ~input return# []]
       (if-let [current# (first input#)]
         (recur (rest input#)
                (let [result# (#'execute-single current#)]
           (if-not (map? current#)
             (conj return# (assoc (#'server-error (var-get #'invalid-request)) "id" nil))
             (if (or (contains? current# "id") (= (get (get result# "error") "code") -32600))
               (conj return# (#'execute-single current#))
               return#))))
       (if-not (empty? (remove nil? return#))
         (json/write-str (vec (remove nil? return#)))
         "")))))

(defmacro do-single
  "Very simply executes a single procedure call. Enforces not returning anything
  if the call is a notification."
  [input]
  `(if (contains? ~input "error")
     (json/write-str ~input)
     (let [result# (#'execute-single ~input)]
       (if (or (contains? ~input "id") (= (get (get result# "error") "code") -32600))
         (json/write-str (#'execute-single ~input))
         ""))))

(defn- execute
  "Handles request strings, dealing with batches etc. along the way."
  [request]
  (let [input (try (json/read-str request)
                   (catch Exception _
                     (assoc (server-error parse-error) "id" nil)))]
    (if (vector? input)
      (do-batch input)
      (do-single input))))

; From http://stackoverflow.com/a/12503724
(defn- parse-int
  "Parses the first continuous number only"
  [s]
  (new Integer (re-find #"\d+" s)))

(defn- socket-io
  "Reads all incoming data from the socket, then sends the completed string to
  execute, before writing the return data to the socket."
  [socket]
  (async/go-loop [length 0 request ""]
    (when-let [line (async/<! (:in socket))]
      (if (zero? length)
        (recur (parse-int line) request)
        (let [request (str request line)]
          (if (< (count request) length)
            (recur length (str request "\n"))
            (let [result (execute request)]
              (let [count (count result)]
                (if (zero? count)
                  (async/>! (:out socket) zero-string)
                  (async/>! (:out socket) (str count "\n" result)))))))))))

(defmacro defprocedure
  "Creates a new procedure that can be called through the JSON-RPC interface.
  The supplied function receives two arguments automatically - 'args', which is
  a vector of unnamed arguments (i.e. the caller supplied an Array), and
  'notification', which is true when the function should not return anything.
  The named-args argument takes a vector of named arguments, and is used if the
  caller supplied a parameters Object. The names must be specified in the same
  order as they should appear in the unnamed arguments version."
  [sym named-args func]
  `(if-not (str/starts-with? (name (quote ~sym)) "rpc.")
     (alter-var-root #'jrpc-namespace
                     (constantly (assoc (var-get #'jrpc-namespace)
                                        (name (quote ~sym))
                                        (list ~named-args
                                              (fn [~'args ~'notification] ~func)))))
     (throw (new IllegalArgumentException
                     "rpc.* method names are reserved for system extensions"))))

(defn start-async
  "Creates a new asynchronous socket server with a given port, optional backlog
  (maximum queue length of incoming connection indications, 50 by default) and
  optional bind address (localhost by default), and returns immediately with a
  communications channel that can be used to command the server to shut down."
  ([port]
    (start-async port default-server-backlog))
  ([port backlog]
    (start-async port backlog nil))
  ([port backlog bind-addr]
    (let [server (socket-server port backlog (InetAddress/getByName bind-addr))]
      (.addShutdownHook (Runtime/getRuntime)
                        (new Thread (fn [] (stop-socket-server server))))
      (let [shutdown (promise)]
        (list
         (async/go-loop []
           (when-let [connection (async/<! (:connections server))]
             (socket-io connection)
             (when-not (realized? shutdown)
               (recur))))
         shutdown)))))

(defn start
  "Creates a new asynchronous socket server with a given port, optional backlog
  (maximum queue length of incoming connection indications, 50 by default) and
  optional bind address (localhost by default), blocking the calling thread in
  the process. Use this server instead of the async server to prevent the main
  thread from exiting."
  ([port]
    (start port default-server-backlog))
  ([port backlog]
    (start port backlog nil))
  ([port backlog bind-addr]
    (let [server (start-async port backlog bind-addr)]
      (loop []
        (when (async/<!! (first server))
          (recur))))))
